pkg/ipam-controller/allocator/allocator.go

```diff
@@ -0,0 +1,393 @@
+
+// contains allocation information for the node
+type nodeAllocationInfo struct {
+	Node    string
+	Subnet  *net.IPNet
+	Gateway net.IP
+	allocatedRange
+}
+
+// allocatedRange contains range of IPs allocated for the node
+type allocatedRange struct {
+	StartIP net.IP
+	EndIP   net.IP
+}
+
+func newPoolAllocator(cfg AllocationConfig) *poolAllocator {
+	return &poolAllocator{
+		cfg:         cfg,
+		allocations: map[string]allocatedRange{}}
+}
+
+// poolAllocator contains pool settings and related allocations
+type poolAllocator struct {
+	cfg AllocationConfig
+	// allocations for nodes, key is the node name, value is allocated range
+	allocations map[string]allocatedRange
+}
+
+func (pa *poolAllocator) getLog(ctx context.Context, cfg AllocationConfig) logr.Logger {
+	return logr.FromContextOrDiscard(ctx).WithName(fmt.Sprintf("allocator/pool=%s", cfg.PoolName))
+}
+
+// Configure update configuration for pool allocator, resets allocations if required
+func (pa *poolAllocator) Configure(ctx context.Context, cfg AllocationConfig) {
+	log := pa.getLog(ctx, cfg)
+	log.V(1).Info("pool configuration update")
+	if pa.cfg.Equal(&cfg) {
+		log.V(1).Info("pool configuration is the same, keep allocations")
+		return
+	}
+	pa.cfg = cfg
+	pa.allocations = map[string]allocatedRange{}
+	log.Info("pool configuration updated, reset allocations")
+}
+
+// Allocate allocates a new range in the poolAllocator or
+// return existing one,
+// returns ErrNoFreeRanges if no free ranges available
+func (pa *poolAllocator) Allocate(ctx context.Context, node string) (nodeAllocationInfo, error) {
+	log := pa.getLog(ctx, pa.cfg).WithValues("node", node)
+	existingAlloc, exist := pa.allocations[node]
+	if exist {
+		log.V(1).Info("allocation for the node already exist",
+			"start", existingAlloc.StartIP, "end", existingAlloc.EndIP)
+		return pa.getNodeAllocationInfo(node, existingAlloc), nil
+	}
+	allocations := pa.getAllocationsAsSlice()
+	var startIP net.IP
+	if len(allocations) == 0 || ip.Distance(ip.Network(pa.cfg.Subnet).IP, allocations[0].StartIP) > 2 {
+		// start allocations from the network address if there are no allocations or if the "hole" exist before
+		// the firs allocation
+		startIP = ip.NextIP(ip.Network(pa.cfg.Subnet).IP)
+	} else {
+		for i := 0; i < len(allocations); i++ {
+			nextI := i + 1
+			// if last allocation in the list
+			if nextI > len(allocations)-1 ||
+				// or we found a "hole" in allocations. the "hole" can't be less than required for the allocation by design.
+				// because we reset all allocations when PerNodeBlockSize size changes
+				ip.Distance(allocations[i].EndIP, allocations[nextI].StartIP) > 1 {
+				startIP = ip.NextIP(allocations[i].EndIP)
+				break
+			}
+		}
+	}
+	nodeRange := allocatedRange{
+		StartIP: startIP,
+		EndIP:   ip.NextIPWithOffset(startIP, int64(pa.cfg.PerNodeBlockSize)-1),
+	}
+	if !pa.cfg.Subnet.Contains(nodeRange.EndIP) || ip.IsBroadcast(nodeRange.EndIP, pa.cfg.Subnet) {
```

Comment:
can you also add check that start IP is in subnet ?

far fetched scenario where startIP has wrap around to 0 and EndIP is somehow within the subnet.